src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,370 @@
+            # Exception ignored in: 'rados.Watch.__dealloc__'
+            # Traceback (most recent call last):
+            #   File "rados.pyx", line 477, in rados.Rados.require_state
+            # rados.RadosStateError: RADOS rados state (You cannot perform that operation on a Rados object in state shutdown.)
+            self._config_watch = None
+        if self._ioctx:
+            self._ioctx.close()
+        self._cluster.shutdown()
+        logger.debug("Cluster connection is shut down")
+
+    def _config_notify(
+        self, notify_id: int, notifier_id: str, watch_id: int, data: bytes
+    ) -> None:
+        key = data.decode("utf-8")
+        logger.debug(
+            f"Got notify on config object {notify_id} {notifier_id} {watch_id} {key}"
+        )
+        if key not in self._watches:
+            return
+
+        value = self._get(key)
+        for watch in self._watches[key].values():
+            watch(key, value)
 
     async def close(self) -> None:
         """Close k/v store connection"""
-        self._is_closing = True
-        if not self._client:
-            self._is_open = False
-            return
-        await self._client.close()
-        self._client = None
-        self._is_open = False
+        self._run = False
+        self._event.set()
 
     async def put(self, key: str, value: str) -> None:
         """Put key/value pair"""
-        assert self._client
-        await self._client.put(key, value)
+        # Try to write to the kvstore in our pool, and also
+        # write to our local cache (whether or not the write
+        # to the pool succeeds)
+        # Or: should we throw an exception if the write fails?
+        # (or gets stuck for too long)?
+        logger.debug(f"Put {key}: {value}")
+        bvalue = value.encode("utf-8")
+        if self._ioctx:
+            try:
+                # Note that current implementation lands in exception handler
+                # with "RADOS rados state (You cannot perform that operation on a
+                # Rados object in state configuring.)" in ... something (probably
+                # the perform op), but we can short-circuit this if we use
+                # the classes ioctx, and check if it's set first, before even
+                # bothering to try anything.  Same in every other method that
+                # does something with an ioctx.
+                # But we probably also need to look at a lock on it (because...
+                # we *know* this is a problem, or we just imagine it's a problem?)
 
-    async def get(self, key: str) -> Optional[str]:
-        """Get value for provided key"""
-        assert self._client
-        value, _ = await self._client.get(key)
+                with rados.WriteOpCtx() as op:
+                    self._ioctx.set_omap(op, (key,), (bvalue,))
+                    logger.debug("Doing write op")
+                    # write op naturally gets stuck if cluster badly degraded,
```

Comment:
1) Wondering if there is a way to hack the pg placement somehow to have more control :) Thinking out loud here, but would it make sense to scale the pool size to the number of aquarium nodes? 
Using the upmap feature[42] to pin the pg where we want it to be? As this map is quite important, perhaps even on every node?

2) Probably a naive question, but would it make sense to put the data into the cluster map or the Ceph config store? It would have the same availability as the mons, but updating is probably more expensive.

[42] https://docs.ceph.com/en/latest/rados/operations/balancer/#modes

Reply:
> 2\. Probably a naive question, but would it make sense to put the data into the cluster map or the Ceph config store? It would have the same availability as the mons, but updating is probably more expensive.

We previously discussed storing this stuff in the ceph config store. The only written reference I can find in a hurry is https://github.com/aquarist-labs/aquarium/issues/32, which is pretty terse. IIRC arguments against were:
* Practical: We can't access the data if the cluster is down, but of course the local cache fixes (or should fix) that, which kinda nullifies this point.
* Aesthetic: The ceph config store is for storing _ceph_ configuration, and aquarium is somehow "outside" ceph, or ceph is inside aquarium ;-) because aquarium is deploying it. From that perspective, we'd be "polluting" the ceph config store
* I can't remember if there were other concerns :-/ (@jecluis? @mgfritch?)

Reply:
> 1. Wondering if there is a way to hack the pg placement somehow to have more control :) Thinking out loud here, but would it make sense to scale the pool size to the number of aquarium nodes?
Using the upmap feature[42] to pin the pg where we want it to be? As this map is quite important, perhaps even on every node?

This would be overkill. It would be fine for a cluster up to 3 nodes or so, but then it becomes a liability because the pool would be degraded should a node die without anywhere to move the data to. Also, depending on what data we're storing, should we start storing lots of small writes (e.g., prometheus data), a single update would have to it all nodes in the cluster.

> I can't remember if there were other concerns :-/ (@jecluis? @mgfritch?)

Lack of watches, for one. Also, it would be nice if we could persist prometheus data in this pool. That way we don't need to have multiple instances of prometheus running, just the one and rely on Ceph for redundancy and persistence.